cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaRebalanceAssemblyOperator.java

```diff
@@ -733,7 +733,7 @@ private Future<MapAndStatus<ConfigMap, KafkaRebalanceStatus>> onPendingProposal(
 
         if (rebalanceAnnotation(kafkaRebalance) == KafkaRebalanceAnnotation.refresh) {
             LOGGER.infoCr(reconciliation, "Requesting a new proposal since refresh annotation is applied on the KafkaRebalance resource");
-            requestRebalance(reconciliation, host, apiClient, kafkaRebalance, true, rebalanceOptionsBuilder).onSuccess(p::complete);
+            return requestRebalance(reconciliation, host, apiClient, kafkaRebalance, true, rebalanceOptionsBuilder);
```

Comment:
I think this works. But I'm not sure this is a good design. You are mixing returning the method results directly with using the promise and returning it at the end. Would it make sense to use a single way at least in the whole method? I think this just increases the risk of more issues being added there in the future because of these various implementations.

In general, using only returning of the method results helps because IDE checks the returns. But even if we use the promise everywhere I think the consistency without the IDE checks would be still better.

(Same applies below.)

Reply:
The first implementation I had was exactly the way you are commenting, by completing/failing the promise within onSuccess and onFailure called on the Future returned by the requestRebalance method.
Then I realised that it was just a "duplication" of what happens when we just return the Future of such a method.
Other onXXX methods (handling other states of the rebalancing FSM) are just doing the return with no additional promise so I went this way, where possible, while leaving the promise where it was needed because of more "nested" async calls.
I will think more to have more consistency.

Reply:
If you don't want to use the promise and just return I think you can use for example `recover` or `eventually` if that is what you meant.

Reply:
I decided to go the promise way. There are anyway two other places where it's needed.